host/crypto/openssl/cert.c

```diff
@@ -294,49 +274,108 @@ static STACK_OF(X509) * _clone_chain(STACK_OF(X509) * chain)
+    X509_STORE* store = NULL;
+    X509* x509 = NULL;
     STACK_OF(X509)* chain = NULL;
 
     /* Clone the certificate to clear any cached verification state */
-    if (!(cert = _clone_x509(cert_)))
-        OE_RAISE(OE_FAILURE);
+    if (!(x509 = _clone_x509(cert)))
+        OE_RAISE_MSG(OE_FAILURE, "Failed to clone X509 cert", NULL);
 
     /* Clone the chain to clear any cached verification state */
     if (!(chain = _clone_chain(chain_)))
-        OE_RAISE(OE_FAILURE);
+        OE_RAISE_MSG(OE_FAILURE, "Failed to clone X509 cert chain", NULL);
+
+    /* Create a store for the verification */
+    if (!(store = X509_STORE_new()))
+        OE_RAISE_MSG(OE_FAILURE, "Failed to allocate X509 store", NULL);
 
     /* Create a context for verification */
     if (!(ctx = X509_STORE_CTX_new()))
-        OE_RAISE(OE_FAILURE);
+        OE_RAISE_MSG(OE_FAILURE, "Failed to create new X509 context", NULL);
 
     /* Initialize the context that will be used to verify the certificate */
-    if (!X509_STORE_CTX_init(ctx, NULL, NULL, NULL))
-        OE_RAISE(OE_FAILURE);
+    if (!X509_STORE_CTX_init(ctx, store, NULL, NULL))
+    {
+        OE_RAISE_MSG(OE_FAILURE, "Failed to initialize X509 context", NULL);
+    }
+
+    /* Create a store with CRLs if needed */
+    if (crls && num_crls)
+    {
+        X509_VERIFY_PARAM* verify_param = NULL;
+
+        for (size_t i = 0; i < num_crls; i++)
+        {
+            crl_t* crl_impl = (crl_t*)crls[i];
+
+            /* X509_STORE_add_crl manages its own addition refcount */
+            if (!X509_STORE_add_crl(store, crl_impl->crl))
+                OE_RAISE_MSG(
+                    OE_FAILURE, "Failed to add CRL to X509 store", NULL);
+        }
+
+        /* Get the verify parameter (must not be null) */
+        if (!(verify_param = X509_STORE_CTX_get0_param(ctx)))
+            OE_RAISE_MSG(
+                OE_FAILURE, "Failed to get X509 verify parameter", NULL);
+
+        X509_VERIFY_PARAM_set_flags(verify_param, X509_V_FLAG_CRL_CHECK);
+        X509_VERIFY_PARAM_set_flags(verify_param, X509_V_FLAG_CRL_CHECK_ALL);
+    }
 
     /* Inject the certificate into the verification context */
-    X509_STORE_CTX_set_cert(ctx, cert);
+    X509_STORE_CTX_set_cert(ctx, x509);
 
     /* Set the CA chain into the verification context */
     X509_STORE_CTX_trusted_stack(ctx, chain);
 
     /* Finally verify the certificate */
     if (!X509_verify_cert(ctx))
-        OE_RAISE(OE_FAILURE);
+    {
+        oe_result_t verify_result = OE_VERIFY_FAILED;
+        int errorno = X509_STORE_CTX_get_error(ctx);
+        switch (errorno)
+        {
+            case X509_V_ERR_CRL_HAS_EXPIRED:
+                verify_result = OE_VERIFY_CRL_EXPIRED;
+                break;
+            case X509_V_ERR_UNABLE_TO_GET_CRL:
+                verify_result = OE_VERIFY_CRL_MISSING;
+                break;
+            case X509_V_ERR_CERT_REVOKED:
+                verify_result = OE_VERIFY_REVOKED;
+                break;
```

Comment:
Should there be a default case for **OE_VERIFY_FAILED**?

Reply:
I had one originally but took it out because all the switch block does is remap `OE_VERIFY_FAILED` to a more specific error and the default case was just a reassignment of `OE_VERIFY_FAILED ` to the internal `verify_result` that was already initialized to it.

Reply:
Sounds fine.